### ISSUE TITLE
Create installation scripts

### ISSUE THREAD

**Issue by nguyenLisa** [ABenchM/abm#17]
To facilitate the installation of ABM on a Linux machine.

**Comment by ankur-1989** [comment]
Script is ready , doing final testing of it.

**Comment by nguyenLisa** [comment]
The script does not work on my machine (Ubuntu on VirtualBox). The first time it runs, it fails to install the Docker containers. I logged out and logged it again, then relaunched the script.
After the second run, I tried to launch Eclipse but it failed with a "There was an error launching the application".

**Comment by ankur-1989** [comment]
I just followed the steps to reproduce on my machine and did not get any error and tested everything after running the installation script. and added one more thing to generate the generic hermes configuration directory and transfer the hermes configuration files there. committed the latest script on github master branch.

**Comment by ankur-1989** [comment]
Syntax to call the script 
HERMES_CONFIG_DIR=anypath ./abm_install 

 if we dont pass the HERMES_CONFIG_DIR it will take /opt/abm by default

**Comment by nguyenLisa** [comment]
I ran the newer version of the script. Eclipse installed correctly, however:
- The pull was not done correctly: the abm repository was nowhere to be found
- The only docker image running is the Hermes one. Maven and SBT are not showing in "docker images"
- The user does not have the rights to run docker commands, even after logging out and in again

**Comment by bhermann** [comment]
I tried on a plain Debian 9.3.0 in newly-installed VM. Here are a few things I notices:

- The script requires and already installed git and sudo. This could be checked and installed if not present.
- The script requires that the current user has sudo rights, but does not stop if this is not the case. Thus, parts are executed and parts are left out. 
- If get an error for line 3: `abm_install.sh: Zeile 3: =/opt/abm: Datei oder Verzeichnis nicht gefunden`
- The installation of mysql could not be completed. I get asked for the root password, but it doesn't accept the root password. I get the following:
```
NOTE: RUNNING ALL PARTS OF THIS SCRIPT IS RECOMMENDED FOR ALL MariaDB
      SERVERS IN PRODUCTION USE!  PLEASE READ EACH STEP CAREFULLY!

In order to log into MariaDB to secure it, we'll need the current
password for the root user.  If you've just installed MariaDB, and
you haven't set the root password yet, the password will be blank,
so you should just press enter here.

Enter current password for root (enter for none): 
ERROR 1698 (28000): Access denied for user 'root'@'localhost'
Enter current password for root (enter for none): 
ERROR 1698 (28000): Access denied for user 'root'@'localhost'
Enter current password for root (enter for none): 
ERROR 1698 (28000): Access denied for user 'root'@'localhost'
```
- I get an error for line 33: `abm_install.sh: Zeile 33: mysql_config_editor: Kommando nicht gefunden.
mysql: unknown variable 'login-path=abm'`

- The apt-get install command sometimes ask for user input. It might be a good idea to provide the -y option.
- The installation of docker failed, but that might be a Debian apt issue, because the https transport is not installed as default (which is kind of offputting btw). 
```
E: Der Treiber für Methode /usr/lib/apt/methods/https konnte nicht gefunden werden.
N: Ist das Paket apt-transport-https installiert?
E: Fehlschlag beim Holen von https://download.docker.com/linux/ubuntu/dists/stretch/InRelease  
```

**Comment by ankur-1989** [comment]
Noted!! Working on it.

**Comment by ankur-1989** [comment]
Script is completed now and have updated on github with abm_install.zip under docs folder.
Steps to follow :- 
 1. Download the zip file and unzip it.
 2.  It contains abm_install.sh which we need to execute.
 3. Command to run the script - 

sudo  sh mailid=your_email_id user=your_local_user MYSQL_ROOT_PASSWORD=your_choice_password HERMES_CONFIG_DIR=your_choice_dir abm_install.sh

MYSQL_ROOT_PASSWORD is optional . it will take password "password" by default. Criteria to set the password is 
# SET THIS! Avoid quotes/apostrophes in the password, but do use lowercase + uppercase + numbers + special chars

HERMES_CONFIG_DIR is optional . It will take /opt/abm by default.

mailid is also optional and commented in the script as of now. If you want final logs on ur email after installed finished then provide your email id and you will have to configure mailx utility on your machine first. or any other known utility. 

user :- it is non root user on your machine through which application will run. It is mandatory to pass.By default it takes abm. 

I have changed lot of things after Ben's feedback. It should work now.

**Comment by ankur-1989** [comment]
one more thing, now the script will do completely silently install will not ask for any input.

**Comment by ankur-1989** [comment]
once the installation is done , post installation steps are 

1. Install bndtools in Eclipse
2. Import project in Eclipse
3.  Run the application and configure the demo user as per the instructions in developer_manual. 

I will update all the instructions in Developer_manual as well once this is fully tested by others.

**Comment by nguyenLisa** [comment]
Thanks Ankur. I tested the new script, and here's my feedback.
- The command should have the script file earlier: `sudo sh abm_install.sh mailid=your_email_id...`
- Could you clean up the docs folder by removing the old scripts?
- Error in the installation of docker-ce.
- Script stuck at the sql installation.
Terminal prints below:
```
Cloning into 'abm'...
remote: Counting objects: 3719, done.
remote: Compressing objects: 100% (47/47), done.
remote: Total 3719 (delta 32), reused 23 (delta 11), pack-reused 3655
Receiving objects: 100% (3719/3719), 55.13 MiB | 3.20 MiB/s, done.
Resolving deltas: 100% (1702/1702), done.
Checking connectivity... done.
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Package docker-ce is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source

E: Package 'docker-ce' has no installation candidate
abm_install.sh: 39: abm_install.sh: docker: not found
abm_install.sh: 41: abm_install.sh: docker: not found
removed '/home/abm/secure_our_mysql.sh'
mysql: [Warning] Using a password on the command line interface can be insecure.
mysql: [Warning] Using a password on the command line interface can be insecure.
mysql: [Warning] Using a password on the command line interface can be insecure.
```

**Comment by ankur-1989** [comment]
From the logs i can see apt-get repository does not have docker packages which I found for myself but I was expecting it. This I will change. Second mysql installation is silent so it remains there for some time and then prints log so it feels it is stucked, but through logs it does not look like it failed.

**Comment by ankur-1989** [comment]
Old script deleted from the docs folder.

**Comment by nguyenLisa** [comment]
Ok, let me know when you have the new script and I'll test it.
How long does the sql installation typically takes? I let it run for 10 min roughly before stopping it.

**Comment by ankur-1989** [comment]
it should not take 10 mins . even it should not stuck because there is no such command which depends on something.I'll look into it and will update soon.

**Comment by nguyenLisa** [comment]
Thanks!

**Comment by ankur-1989** [comment]
So I found the issue. Script was getting stuck at java installation. I missed the -y option during that and it was waiting for Y input but does not show on console as I was logging all the console output in log file.  Fixed this and docker part too. Tested on fresh Ubuntu image and working perfectly. attaching the logs. 

You have to log out and log in after the installation.

**Comment by ankur-1989** [comment]
or better to restart sometimes logout login does not work.

**Comment by ankur-1989** [comment]
[2018-02-16 17:59:30 root.log](https://github.com/ABenchM/abm/files/1733130/2018-02-16.17.59.30.root.log)
[2018-02-16 18:14:47 root.log](https://github.com/ABenchM/abm/files/1733131/2018-02-16.18.14.47.root.log)

**Comment by nguyenLisa** [comment]
Tested until the script finished. The installation seems to work fine. Here's my feedback:
- [ ] The Eclipse version that is installed is 3.8 (Juno or Indigo if I am not mistaken). Would it be possible to install a newer version?
- [ ] The mailing command is commented out. It would be nice to have it in again. Or was it commented out because it caused errors? In this case, we can leave it commented out.
- [ ] Building sbt and the part of the script after setting up the database take a long time and could be interpreted as a stuck script. Perhaps more prints would help.
- [ ] The HERMES_CONFIG_DIR option does not work. The hermes files are copied directly under /opt/abm.

Thanks for the work so far. Could you let me know when you have a new version so that I can test the full installation?

**Comment by ankur-1989** [comment]
1. Yes in installs old eclipse which it finds latest in Ubuntu repository. To install the latest version , either update the repository or download the latest version. 
2. Mailx utility will work , user will have to configure the mailx utility in his or her system. i did not keep it open because there are multiple mail agents and different requirements and it asks for particular configurations. so I left it commented.
3. Yeah it takes in installing all the images but we are maintaining log file in $HOME location so we can keep tracking the log file for the updates and this was kept like this intentionally.
4. however it should work , i can check this part.
5. java and eclipse logs missing but as you said you saw the eclipse version it means script ran completely.

**Comment by nguyenLisa** [comment]
1. Yes, please change that.
2. Then, please remove the mailing option from the commandline. 
3. Cool, thanks. Follow-up: at the beginning of the script, please print the path to the log files, just in case.

Let me know when you have an updated version with a newer Eclipse. I'll test it then. Thanks!

**Comment by ankur-1989** [comment]
Uploaded the file with the latest changes now. let me know if any issues found.

**Comment by nguyenLisa** [comment]
Thanks, the scripts work. I did a few modifications to simplify the commandline options and beautify what we see in the terminal. I tested it on an Ubuntu VM, and everything works (installation, set-up of Eclipse and the ABM local configurations, creation of a demo user, creation of a collection (containing the project apache commons fileupload), build of a collection and Hermes run on that collection).

Could you check if it works for you? (see documentation here https://github.com/ABenchM/abm/blob/master/docs/Developer_Manual.md#installation-scripts-linux-only)

**Comment by ankur-1989** [comment]
Lisa I noticed one thing in the changes that you have added bash in the script in front of commands. Will it work for other shell kernels like sh , ksh or others ?

**Comment by ankur-1989** [comment]
I'll check the changes.

**Comment by ankur-1989** [comment]
Yeah , Script is working fine. Shall I close this ticket now ?

**Comment by nguyenLisa** [comment]
Sure!
